### DISCUSSION TITLE
Why doesn't set_stopped allow a value to be passed like set_value/set_error?

### DISCUSSION THREAD

**Discussion by bnbajwa** [NVIDIA/stdexec#1019] [Q&A] [answered]
The 'set_stopped' completion function doesn't allow a value to be passed in unlike set_value/set_error.

In my case, I would like to pass a reason to 'set_stopped' completion function.  io_uring has a multi-shot receive operation, where you make a single request, and you get multiple outputs. However, the request can expire at any time. When it does expire, I want to invoke 'set_stopped' and pass in some value representing the fact that the request stopped on its own.

There is also some other data I want to pass back to set_stopped. Currently, i'm just using set_error for both cancellations/errors, and differentiating which it is based on the type of the value passed back.

What do you think? I've only just started experimenting with this library/P2300 so not sure if there is a way to achieve what I want.

**Comment by maikel** [comment, marked as answer]
Does your sender complete when the whole multi-shot operation completes? In my understanding, you would complete with `set_stopped()` when an outer cancelation request happens. I would write an algorithm that completes with some "partial sum" result when the multi-shot op has stopped or needs to be rescheduled.

If you need to expose each single read operation, using something like a sequence-sender might be more appropriate.

**Reply by maikel** [reply]
To clarify: I would send the partial sum (maybe with `error_code`) via a `set_value` completion

**Reply by bnbajwa** [reply]
Hi, thanks for the response.

The approach I've taken with io_uring multi-shot ops:
- Create a sender and allocate an operation state by connecting it with a receiver
- For each output of the multi-shot op, I invoke set_value() on the SAME receiver (operation state is kept alive)
- When set_stopped/set_error is invoked, the multi-shot op is done and no more set_value are to be invoked

Not sure if it's a correct way of doing it, in very early stages at the moment, have not yet done a test run.

Thanks for  your partial sum suggestion (though, I would prefer not to have any error_code in set_value completion). My current approach is similar to your suggestion:
- I first invoke set_value with the "partial" result
- then I invoke set_stopped

However, I hit an issue with the second point because I couldn't pass in any data to set_stopped. I will just use set_error for both cancellations/errors I guess since I can pass data through it.

**Reply by maikel** [reply]
If, what you want, is to have a "completion" for each single CQE you really want to use a `set_next`-receiver. Instead of calling `set_value` you could call `set_next` with your result which indicates the progress of an async operation.

Note, that `set_next` is not part of P2300, but an extension, which is developed in stdexec. You can find it under [exec/sequence_senders.hpp](https://github.com/NVIDIA/stdexec/blob/main/include/exec/sequence_senders.hpp) and here is an example of [`iterate(range)`](https://github.com/NVIDIA/stdexec/blob/main/include/exec/sequence/iterate.hpp) which "completes" for each item of a range.

In the case of `io_uring` multi-shot operations, it's not completely clear to me when to construct the operation states for each item or how to bound the total number of active ones.

**Comment by dietmarkuehl** [comment]
Completing any receiver more than once with set_value is certainly wrong! In general receivers can and do expect that there is only one call to any of their 3 completion signatures.You may want to have a look at set_next: you’d report partial tesults ising set_next and upon completion you’d complete with set_value, possibly not providing another value rather nothing or some sort of report.On 9 Aug 2023, at 23:02, Baber Nawaz ***@***.***> wrote:﻿
Hi, thanks for the response.
The approach I've taken with io_uring multi-shot ops:

Create a sender and allocate an operation state by connecting it with a receiver
For each output of the multi-shot op, I invoke set_value() on the SAME receiver (operation state is kept alive)
When set_stopped/set_error is invoked, the multi-shot op is done and no more set_value are to be invoked

Not sure if it's a correct way of doing it, in very early stages at the moment, have not yet done a test run.
Thanks for  your partial sum suggestion (though, I would prefer not to have any error_code in set_value completion). My current approach is basically same or similar to your suggestion:

I first invoke set_value with the "partial" result
then I invoke set_stopped

However, I hit an issue with the second point because I couldn't pass in any data to set_stopped. I will just use set_error for both cancellations/errors I guess since I can pass data through it.

**Reply by bnbajwa** [reply]
Thanks for the response, much appreciated! I will have to reconsider my approach for multi-shot operations.
